src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -5619,56 +5598,75 @@ public void setServerPreparedStatementDiscardThreshold(int value) {
         this.serverPreparedStatementDiscardThreshold = Math.max(0, value);
     }
 
+    final boolean isPreparedStatementUnprepareBatchingEnabled() {
+    	return 1 < getServerPreparedStatementDiscardThreshold();
+    }
+
     /**
      * Cleans-up discarded prepared statement handles on the server using batched un-prepare actions if the batching threshold has been reached.
      * 
      * @param force 
      *      When force is set to true we ignore the current threshold for if the discard actions should run and run them anyway.
      */
-    final void handlePreparedStatementDiscardActions(boolean force) {
-        // Skip out if session is unavailable to adhere to previous non-batched behavior.
-        if (isSessionUnAvailable()) 
+    final void unprepareUnreferencedPreparedStatementHandles(boolean force) {
+        // Skip out if session is unavailable to adhere to previous non-batched behavior
+        if (isSessionUnAvailable())
             return;
 
+        // Skip out if we are not forced and:
+        //   * another thread is already doing the work of un-preparing, OR
+        //   * number of handles in the map have not crossed the threshold (optimization)
         final int threshold = getServerPreparedStatementDiscardThreshold();
+        if (!force && (isUnprepareInProgress.get() || threshold >= preparedStatementHandleMap.size()))
```

Comment:
@TobiasSQL The only issue with a queue is that a handle is either in or out.  I'm not sure how the semantics work in that case.  In theory, a handle that reaches refcount 0 is reclaimable, but that is the case immediately after a statement is executed.  Before, the tie to the metadata cache, and the eviction, artifically kept handles alive.  But if that relationship is severed, as discussed above, claiming handles as soon as they reach refcount 0 seems inefficient.

In the case of the concurrent collection, a handle that as reached refcount 0 always has the possibility of being "rescued" by a new statement execution -- but only because it remains in the live collection.  In the case of a true unprepare queue, once a handle is enqueued, its fate is sealed.

Now, maybe something like an eviction-based collection, dedicated to handles (separate from the metadata) *could* work, if the behavior is LRU in nature...

Thoughts?